tap_mailchimp/streams.py

```diff
@@ -0,0 +1,538 @@
+
+    def get_batch_info(self, batch_id):
+        """Function to get batch status"""
+        try:
+            return self.client.get('/batches/{}'.format(batch_id), endpoint='get_batch_info')
+        except HTTPError as e:
+            if e.response.status_code == 404:
+                raise BatchExpiredError('Batch {} expired'.format(batch_id))
+            raise e
+
+    def check_and_resume_email_activity_batch(self):
+        """Function to resume batch syncing from previous sync"""
+        batch_id = self.get_bookmark(['reports_email_activity_last_run_id'], None)
+
+        if batch_id:
+            try:
+                data = self.get_batch_info(batch_id)
+                if data['status'] == 'finished' and not data['response_body_url']:
+                    LOGGER.info('reports_email_activity - Previous run from state (%s) is empty, retrying.',
+                                batch_id)
+                    return
+            except BatchExpiredError:
+                LOGGER.info('reports_email_activity - Previous run from state expired: %s',
+                            batch_id)
+                return
+
+            # Resume from bookmarked job_id, then if completed, issue a new batch for processing.
+            campaigns = []  # Don't need a list of campaigns if resuming
+            self.sync_email_activities(campaigns, batch_id)
+
+    def poll_email_activity(self, batch_id):
+        """Get the email activity URL after the batch as executed"""
+        sleep = 0
+        start_time = time.time()
+        while True:
+            data = self.get_batch_info(batch_id)
+
+            # Needs to update frequently for target-stitch to capture state
+            self.write_activity_batch_bookmark(batch_id)
+
+            progress = ''
+            if data['total_operations'] > 0:
+                progress = ' ({}/{} {:.2f}%)'.format(
+                    data['finished_operations'],
+                    data['total_operations'],
+                    (data['finished_operations'] / data['total_operations']) * 100.0)
+
+            LOGGER.info('reports_email_activity - Job polling: %s - %s%s',
+                        data['id'],
+                        data['status'],
+                        progress)
+
+            if data['status'] == 'finished':
+                return data
+            elif (time.time() - start_time) > MAX_RETRY_ELAPSED_TIME:
+                message = 'Mailchimp campaigns export is still in progress after {} seconds. \
+                    Will continue with this export on the next sync.'.format(MAX_RETRY_ELAPSED_TIME)
+                LOGGER.error(message)
+                raise Exception(message)
+
+            sleep = next_sleep_interval(sleep)
+            LOGGER.info('campaigns - status: %s, sleeping for %s seconds',
+                        data['status'],
+                        sleep)
+            time.sleep(sleep)
+
+    def stream_email_activity(self, archive_url):
```

Comment:
Unit test of this method is missing

Reply:
While writing unittest for stream_email_activity function, when trying to mock **tarfile.open()**, the process is getting stuck for a long time and thereby the process is getting killed. 
This is happening because we have mocked the call to **self.client.request()** function and we are providing a test response as its return_value. So the **response.raw** passed as an argument to **tarfile.open()** is taking an infinite time to get processed which is resulting in the process getting killed.
Hence, we have skipped the unittest for this function.